### PR TITLE
#163445913 implementation logic feedback

### PR DESCRIPTION
**Pull request by openwell**
#### What does this PR do?
- This PR integrates improved logic for better performance 
#### Description of Task to be completed?
-  Accessing of API end points
#### How should this be manually tested?
- Pull this branch [ch-implementation-logic-feedback-163445913](https://github.com/openwell/questioner/tree/ch-implementation-logic-feedback-163445913) off this repository
- On your command line, run  `npm install` to install all dependencies
- Run `npm database` to create the database tables and create an admin
- An admin token is generated if successfully and will be displayed in the terminal
- The token should be part of the headers as 'tokens': 'token-code'
- Run `npm start` to start the app
- The app would run on port 3000
- Access the endpoint on `localhost:3000/api/v1/`
#### Any background context you want to provide?
- This app is a node app and would require nodejs and npm installation on your local machine
- This app will require PostgreSQL Database
#### What are the relevant pivotal tracker stories?
[#163445913](https://www.pivotaltracker.com/story/show/163445913)
#### Screenshots (if appropriate)
None
#### Questions:
None

### CODE REVIEW

**Comment by megame24** [comment]
Why do you have a merge commit on this PR?

**Comment by openwell** [comment]
I don't have a merge commit, and my develop branch can show no recent commit was made.
The last action I took was to resolve conflict and git created that commit with that name because the commit itself shows little changes when you click on changes.

**Comment by megame24** [comment]
> I don't have a merge commit, and my develop branch can show no recent commit was made.
> The last action I took was to resolve conflict and git created that commit with that name because the commit itself shows little changes when you click on changes.

Ok, fix your failing tests.

File: src/controller/helpers.js [hunk 1]
```diff
@@ -14,38 +14,55 @@ class auth {
...
 
-  static async verifyToken(req, res, next) {
-    auth.tokenHandler('users', req, res, next);
-  }
-
   static async adminVerifyToken(req, res, next) {
-    auth.tokenHandler('admins', req, res, next);
+    const token = req.headers.tokens;
+    if (!token) {
+      return errorHandler(400, res, 'Token is not provided');
```
Comment (megame24) [inline src/controller/helpers.js]: Do you think that error code 400 is appropriate for this scenario?

File: src/controller/helpers.js [hunk 2]
```diff
@@ -14,38 +14,55 @@ class auth {
...
+        return errorHandler(403, res, 'Token Forbidden');
+      }
+      const { rows } = await db.query(queries.selectById('admins', 'id', decoded.userId));
+      if (!rows[0]) {
+        return errorHandler(401, res, 'Invalid token');
+      }
+      req.user = { id: decoded.userId, isadmin: decoded.isAdmin };
+      return next();
+    } catch (error) {
+      return errorHandler(400, res, error);
```
Comment (megame24) [inline src/controller/helpers.js]: In what situations will this errorHandler be called?

File: src/controller/meetUpController.js [hunk 3]
```diff
@@ -128,11 +136,10 @@ class Controller {
...
       const { rows } = await db.query(queries.selectById('questions', 'id', req.params.questionId));
-      const increase = [(rows[0].votes + 1), rows[0].id];
-      const query2 = 'UPDATE questions SET votes = $1 WHERE id = $2 RETURNING meetup_id,title,body, votes';
-      const resp = await db.query(query2, increase);
-      return res.status(200).json({
-        status: 200,
+      const resp = await db.query(queries.updateVote((rows[0].votes + 1), rows[0].id));
+      const { rowCount } = await db.query(queries.votes(req.params.questionId, req.user.id));
+      return res.status(201).json({
+        status: 201,
```
Comment (megame24) [inline src/controller/meetUpController.js]: When should status code 201 be returned and do you think it is appropriate to return 201 in this situation?

File: src/controller/meetUpController.js [hunk 4]
```diff
@@ -144,11 +151,10 @@ class Controller {
...
       const { rows } = await db.query(queries.selectById('questions', 'id', req.params.questionId));
-      const increase = [(rows[0].votes - 1), rows[0].id];
-      const query2 = 'UPDATE questions SET votes = $1 WHERE id = $2 RETURNING meetup_id,title,body, votes';
-      const resp = await db.query(query2, increase);
-      return res.status(200).json({
-        status: 200,
+      const resp = await db.query(queries.updateVote((rows[0].votes - 1), rows[0].id));
+      const { rowCount } = await db.query(queries.votes(req.params.questionId, req.user.id));
+      return res.status(201).json({
+        status: 201,
```
Comment (megame24) [inline src/controller/meetUpController.js]: When should status code 201 be returned and do you think it is appropriate to return 201 in this situation?

File: src/controller/helpers.js (hunk 1 above)
Comment (openwell) [reply src/controller/helpers.js]: I believe it should have been 401, authentication error

File: src/controller/meetUpController.js (hunk 3 above)
Comment (openwell) [reply src/controller/meetUpController.js]: Based on further reading i believe it should be 200

File: src/controller/meetUpController.js (hunk 4 above)
Comment (openwell) [reply src/controller/meetUpController.js]: Based on further reading i believe it should be 200

File: src/controller/helpers.js (hunk 2 above)
Comment (openwell) [reply src/controller/helpers.js]: In a case where there is server error or database error

Comment (openwell) [reply src/controller/helpers.js]: In a case where there is server error or database error

File: src/controller/helpers.js (hunk 1 above)
Comment (megame24) [reply src/controller/helpers.js]: Take note and incorporate it where ever it applies.

File: src/controller/helpers.js (hunk 2 above)
Comment (megame24) [reply src/controller/helpers.js]: Ok, what does status code 400 mean, and do you think it applies to this case?

File: src/controller/meetUpController.js (hunk 4 above)
Comment (megame24) [reply src/controller/meetUpController.js]: Please take note of this and incorporate it where ever it applies.

File: src/controller/helpers.js (hunk 2 above)
Comment (openwell) [reply src/controller/helpers.js]: No it should be a 500 error code

File: src/test/allTest.js [hunk 5]
```diff
@@ -256,73 +250,58 @@ describe('/POST /api/v1/comments', () => {
...
 describe('/POST /api/v1/comments', () => {
-  it('it should return 400 Error', (done) => {
+  it('it should return 404 Error', (done) => {
     request(server)
       .post('/api/v1/comments')
       .set('tokens', userToken)
       .send(comment2)
       .end((err, res) => {
-        res.should.have.a.status(400);
+        res.should.have.a.status(404);
```
Comment (megame24) [inline src/test/allTest.js]: You should be testing for more than the status code, I recommend you also test that error message is what it is meant to be. Effect this change where ever it is needed.

Comment (openwell) [reply src/test/allTest.js]: Alright, Working on it

File: .evn-sample [hunk 6]
```diff
@@ -0,0 +1,5 @@
+DATABASE_URL=postgres://{db_username}:{db_password}@{host}:{port}/{db_name}
```
Comment (akinwale-habibullah) [inline .evn-sample]: Good work. There's a little typo with your file name. Use sample.env rather than evn.sample

Comment (akinwale-habibullah) [reply .evn-sample]: Also remove the curly braces, as this can be misleading at times.

File: README.md [hunk 7]
```diff
@@ -3,55 +3,61 @@
...
-or bottom of the log..
+Questioner helps the meetup organizer prioritize
+questions to be answered.
 
 ## Getting Started
 
 Clone the Repo.
 -------------
-Run npm install to install all necessary packages.
+Run `npm install` to install all necessary packages.
```
Comment (akinwale-habibullah) [inline README.md]: I think you should add these steps;

- Rename sample.env file
- Add the required env variables in the sample.env file
- Create database with a given name (questioner)

File: README.md [hunk 8]
```diff
@@ -3,55 +3,61 @@
...
+- DELETE `/api/v1/meetups/:meetup-id` Delete a meetup.
+
 
 ### Installing
 
 ## On your Local Machine
-- Pull the [develop](https://github.com/openwell/questioner) branch off this repository
+- Pull the [develop](https://github.com/openwell/Questioner) branch off this repository
+- Install postgreSql 11 on your local machine or registered on any online platform
+- create a .env file
```
Comment (akinwale-habibullah) [inline README.md]: Why not make this rename sample.env file to .env and input all the required environment variables.

**Review by akinwale-habibullah — APPROVED** [review]
Kudos. I think you did a great job. Your app is quite robust. A few things you need to work on before you merge:

- In index.js please rename the error middleware you imported from errorHandler.js file
- Add some eslint rules as we discusssed.
- Return the appropriate HTTP status code when there no meetups, questions and comments records instead of 404 a user gets when you do a GET to these resources
- Return a descriptive response when admin logs in rather than http 500 status code.

**Comment by akinwale-habibullah** [comment]
Also look at the codeclimate report, and try to work on them. I think you need to turn offf some ESLint rules.

**Pull request merged**